File: csharp/834.sum-of-distances-in-tree.cs
/*
 * @lc app=leetcode id=834 lang=csharp
 *
 * [834] Sum of Distances in Tree
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {

    private static void BuildRet(IList<int>[] map, int[] count, int[] ret, int N, int prev, int curr) {
        foreach (var next in map[curr]) {
            if (next == prev) {
                continue;
            }
            ret[next] = N - 2 * count[next] + ret[curr];
            BuildRet(map, count, ret, N, curr, next);
        }
    }

    private static void BuildCount(IList<int>[] map, int[] count, int[] ret, int prev, int curr) {
        count[curr] = 1;
        foreach (var next in map[curr]) {
            if (next == prev) {
                continue;
            }
            BuildCount(map, count, ret, curr, next);
            count[curr] += count[next];
            ret[curr] += count[next] + ret[next];
        }
    }

    public int[] SumOfDistancesInTree(int N, int[][] edges) {
        var map = new IList<int>[N];
        var count = new int[N];
        var ret = new int[N];
        for (int i = 0; i < N; ++i) {
            map[i] = new List<int>();
        }
        foreach (var edge in edges) {
            int x = edge[0], y = edge[1];
            map[x].Add(y);
            map[y].Add(x);
        }
        BuildCount(map, count, ret, -1, 0);
        BuildRet(map, count, ret, N, -1, 0);
        return ret;
    }
}
// @lc code=end

File: java/834.sum-of-distances-in-tree.java
/*
 * @lc app=leetcode id=834 lang=java
 *
 * [834] Sum of Distances in Tree
 */

// @lc code=start
class Solution {

    private static void buildCount(List<Integer>[] map, int curr, int prev, int[] count, int[] ret) {
        count[curr] = 1;
        for (Integer next : map[curr]) {
            if (prev == next) {
                continue;
            }
            buildCount(map, next, curr, count, ret);
            count[curr] += count[next];
            ret[curr] += ret[next] + count[next];
        }
    }

    private static void buildRet(List<Integer>[] map, int curr, int prev, int[] count, int[] ret) {
        for (Integer next : map[curr]) {
            if (prev == next) {
                continue;
            }
            ret[next] = ret[curr] + map.length - count[next] - count[next];
            buildRet(map, next, curr, count, ret);
        }
    }

    public int[] sumOfDistancesInTree(int N, int[][] edges) {
        List<Integer>[] map = new List[N];
        int[] count = new int[N], ret = new int[N];
        for (int i = 0; i < N; ++i) {
            map[i] = new ArrayList<Integer>();
        }
        for (int[] edge : edges) {
            map[edge[0]].add(edge[1]);
            map[edge[1]].add(edge[0]);
        }
        buildCount(map, 0, -1, count, ret);
        buildRet(map, 0, -1, count, ret);
        return ret;
    }
}
// @lc code=end

